### PR TITLE
Add protection on namespace cache in case namespace is not created

### PR DESCRIPTION
**Pull request by wxing1292**
**What changed?**
Add additional frequency check on namespace cache to reduce the impact if namespace is not registered.

**Why?**
Currently, if a namespace does not exists within cache, the logic will check if the namespace exists in DB and possibly perform a cache refreshment. If the namespace does not exists in DB, and caller keeps on calling server APIs, the DB will encounter huge API call volume.

**How did you test it?**
Existing tests
```zsh
temporal|namespace-cache-protection ⇒ go test -count=20 -race go.temporal.io/server/common/cache
ok  	go.temporal.io/server/common/cache	89.144s
```

**Potential risks**
N/A

### CODE REVIEW

File: common/cache/namespaceCache.go [hunk 1]
```diff
@@ -65,6 +66,8 @@ const (
 	namespaceCacheInitialSize = 10 * 1024
 	namespaceCacheMaxSize     = 64 * 1024
 	namespaceCacheTTL         = 0 // 0 means infinity
+	// NamespaceCacheMinRefreshInterval namespace cache refresh interval
```
Comment (sergeybykov) [inline common/cache/namespaceCache.go:69]: "Minimun namespace cache refresh interval" rather?

File: common/cache/namespaceCache.go [hunk 2]
```diff
@@ -492,16 +501,34 @@ UpdateLoop:
 	c.cacheByID.Store(newCacheByID)
 	c.cacheNameToID.Store(newCacheNameToID)
 	c.triggerNamespaceChangeCallbackLocked(prevEntries, nextEntries)
+
+	// only update last refresh time when refresh succeeded
+	c.lastRefreshTime.Store(now)
```
Comment (alexshtin) [inline common/cache/namespaceCache.go:506]: Why `lastRefreshTime` is atomic? This func is called when lock is already occurred. I believe it is safe to use regular `time.Time` here.

Comment (wxing1292) [reply common/cache/namespaceCache.go:506]: it is also used here:
https://github.com/temporalio/temporal/pull/824/files#diff-a1dd08cc18ae4897b834f91ace740ef18fcc17653d02e2b3ad59a9cfdf59f55aR515

File: common/cache/namespaceCache.go [hunk 3]
```diff
@@ -492,16 +501,34 @@ UpdateLoop:
...
 	return nil
 }
 
 func (c *namespaceCache) checkNamespaceExists(
 	name string,
 	id string,
-) error {
+) (bool, error) {
```
Comment (alexshtin) [inline common/cache/namespaceCache.go:513]: I think logic of this func should be changed. It is great that it returns `(bool, error)` now but `bool` should mean exist/not exist and error should indicate the error: persistence or throttling. I would return `service.ResourceExhausted` right from here and propagate it up to the caller. Caller should be able to differentiate if it is actual `NotFound` or "you are pinging me to fast".

**Review by alexshtin — CHANGES_REQUESTED** [review]

File: common/cache/namespaceCache.go (hunk 3 above)
Comment (wxing1292) [reply common/cache/namespaceCache.go:513]: > but bool should mean exist/not exist

not really, bool means continue to check the cache & refresh the cache or not

Comment (wxing1292) [reply common/cache/namespaceCache.go:513]: and there is not guarantee the `check DB see if the namespace exists` logic will be exercised within this function

File: common/cache/namespaceCache.go (hunk 1 above)
Comment (wxing1292) [reply common/cache/namespaceCache.go:69]: the naming convention (at least) in this file is `namespaceCache` + var

File: common/cache/namespaceCache.go (hunk 3 above)
Comment (alexshtin) [reply common/cache/namespaceCache.go:513]: ok, then don't add this `bool` at all. Let it return `serviceerror.NotFound` or `serviceerror.ResourceExhausted` which will be propagated to the caller.

File: common/cache/namespaceCache.go (hunk 1 above)
Comment (alexshtin) [reply common/cache/namespaceCache.go:69]: I think @sergeybykov wanted this:
```suggestion
	// NamespaceCacheMinRefreshInterval is a minimun namespace cache refresh interval.
```

File: common/cache/namespaceCache.go (hunk 3 above)
Comment (wxing1292) [reply common/cache/namespaceCache.go:513]: nop, that is not the purpose of this function

the purpose of this function is check whether to continue the path, i.e. check & potentially force refresh the cache.
the bool return value means whether to continue (either the domain exists, or somebody has already checked within short period of time and it is really likely cache can contain the target domain, but no guarantee)

Comment (wxing1292) [reply common/cache/namespaceCache.go:513]: the functionality (function in question) was originally introduced to solve the issue of 
t = 0, cache got refreshed, to be refreshed again at t = 10
t = 1, new domain got registered
t = 2, caller just want to use the domain (from within cache)

the function in question (along with the following logic) will solve the above case

Comment (wxing1292) [reply common/cache/namespaceCache.go:513]: may be the function name should be changed?

File: common/cache/namespaceCache.go (hunk 1 above)
Comment (wxing1292) [reply common/cache/namespaceCache.go]: my bad

File: common/cache/namespaceCache.go (hunk 3 above)
Comment (alexshtin) [reply common/cache/namespaceCache.go:513]: Yes, change the name. If `existsSomething()` returns `bool` it usually means exists/doesn't exist.

Also you return `NotFound` at t=2 now, right? I think `Unavailable` suites better here (client should wait and try again w/o performing extra actions): https://pkg.go.dev/google.golang.org/grpc/codes (comment to `FailedPrecondition Code = 8`).

Comment (wxing1292) [reply common/cache/namespaceCache.go:513]: function name is changed

Unavailable means server fault, while in this case (namespace missing in cache) is client fault.
the only guarantee we have with customer is: if you namespace is in the DB, it is guaranteed to be in the cache after 10s + creation time

Comment (alexshtin) [reply common/cache/namespaceCache.go:513]: I think clients should not know about cache or any internal details. So if the statement is "namespace creation can take up to 10 seconds" then returning `NotFound` within those 10 seconds makes perfect sense.

File: common/cache/namespaceCache.go [hunk 4]
```diff
@@ -492,16 +501,34 @@ UpdateLoop:
...
+	now := c.timeSource.Now()
+	if now.Sub(c.lastRefreshTime.Load().(time.Time)) < NamespaceCacheMinRefreshInterval {
+		return false, nil
+	}
 
+	c.checkLock.Lock()
+	defer c.checkLock.Unlock()
+
+	now = c.timeSource.Now()
```
Comment (alexshtin) [inline common/cache/namespaceCache.go:522]: Do you really need it one more time here?

File: common/cache/namespaceCache.go [hunk 5]
```diff
@@ -563,9 +590,13 @@ func (c *namespaceCache) getNamespace(
 		return c.getNamespaceByID(id, true)
 	}
 
-	if err := c.checkNamespaceExists(name, ""); err != nil {
+	doContinue, err := c.checkAndRefresh(name, "")
```
Comment (alexshtin) [inline common/cache/namespaceCache.go:593]: Probably there is a better name for `doContinue`?

**Review by alexshtin — APPROVED** [review]

File: common/cache/namespaceCache.go (hunk 4 above)
Comment (wxing1292) [reply common/cache/namespaceCache.go:522]: yes, consider 1000 caller with 10 frontend host with namespace not registered.

on average, there will be ~ 100 concurrent call to this cache, so the line 514 line `now := c.timeSource.Now()` maybe called a while back.

**Pull request merged**